src/main/java/life/qbic/openbis/openbisclient/OpenBisClient.java

```diff
@@ -584,14 +607,47 @@ public boolean isUserAdmin(String userID) {
     SearchResult<Person> res = v3.searchPersons(sessionToken, criteria, options);
     for (Person p : res.getObjects()) {
       for (RoleAssignment r : p.getRoleAssignments()) {
-        if (r.getRole().equals(Role.ADMIN)) {
+        if (r.getRole().equals(role) && r.getRoleLevel().equals(level)) {
           return true;
         }
       }
     }
     return false;
   }
 
+  /**
+   * Returns whether a user's user group is assigned a given role at a given role level
+   * 
+   * @param userID the user's id
+   * @param role the openBIS role
+   * @param level the openBIS role level, denoting if the user and their group has that role for the
+   *        instance or just one or more spaces or projects
+   * @return true, if user has that role through their user group, false otherwise
+   */
+  @Override
+  public boolean usersGroupHasRole(String userID, Role role, RoleLevel level) {
+    ensureLoggedIn();
+    AuthorizationGroupSearchCriteria criteria = new AuthorizationGroupSearchCriteria();
+    AuthorizationGroupFetchOptions options = new AuthorizationGroupFetchOptions();
+    options.withRoleAssignments().withAuthorizationGroup().withRoleAssignments();
+    options.withUsers();
+    SearchResult<AuthorizationGroup> searchResult =
+        v3.searchAuthorizationGroups(sessionToken, criteria, options);
+
+    for (AuthorizationGroup group : searchResult.getObjects()) {
+      for (Person person : group.getUsers()) {
+        if (person.getUserId().equals(userID)) {
+          for (RoleAssignment r : group.getRoleAssignments()) {
+            if (r.getRole().equals(role) && r.getRoleLevel().equals(level)) {
+              return true;
+            }
+          }
+        }
+      }
+    }
+    return false;
```

Comment:
three nested loops o.0. I find this hard to read and difficult to debug. 

My suggestions:

1. Maybe introduce some filter methods that make the code more readable
2. Search for the groups where the users belongs to in the first place (search criteria, if API allows)

Reply:
Good point, I could not fetch the groups directly via the role assignments (which should be possible according to the API), but I can certainly try to add a search subcriterium to find only groups of a specific user.

Reply:
there does not seem to be an easy way to search for groups of a specific user. will look into this more tomorrow